### ISSUE TITLE
Ingest: Provide more robust ingest for Excel and CSV

### ISSUE THREAD

**Issue by eaquigley** [IQSS/dataverse#585]
---

Author Name: **Kevin Condon** (@kcondon)
Original Redmine Issue: 4014, https://redmine.hmdc.harvard.edu/issues/4014
Original Date: 2014-05-21
Original Assignee: Eleni Castro

---

The initial Excel ingest implementation is a proof of concept -it supports rectangular columns and rows of string and numeric types.

More functionality may be required, for example dates, missing value support, variable labels, but it is unclear what is typically expected from researchers who use Excel to store their data.

This ticket is a place holder to investigate a concrete set of requirements for a more robust Excel ingest.

As a side note, there are a number of "How to use Excel for Research Data" guides posted by various universities out there that may answer some questions and provide a starting point.

**Comment by scolapasta** [comment]
Probably best to just move this to 4.1, but first checking with @landreev if there are any low hanging fruit here

**Comment by landreev** [comment]
@posixeleni : 
Eleni, I've spoken to Gustavo and Liz about this; we decided that we'll attend to putting together any best practices/guidelines documentation in 4.1. 

In the short run though (for 4.0), any chance you know of any real-life Excel files, that any of our partners would actually want to ingest as tabular data? 

I'll try and run the ingest thingy on more XLSX files from production; but if you have any extra leads - please let me know.

**Comment by posixeleni** [comment]
@landreev unfortunately my leads for getting real-life Excel files from our partners fell through. They never sent me any after several follow-up attempts. Sorry I couldnt be of more assistance. But will send some your way if I ever do get some excel data from in-the-wild.

**Comment by pdurbin** [comment]
I mentioned we're looking for XLSX files [#sciencelab](http://logs.glob.uno/?c=mozilla%23sciencelab&s=30+Jan+2015&e=30+Jan+2015) and @kaythaney from @mozillascience [tweeted about it](https://twitter.com/kaythaney/status/561216182779658240)! Thanks @kaythaney! She also suggested emailing http://lists.software-carpentry.org/mailman/listinfo/discuss_lists.software-carpentry.org which is an organization @devbioinfoguy is involved in, I believe.

**Comment by pbulsink** [comment]
Any specific type of excel files you're looking for? I'm sure I've got (anonymized) UV-Vis data (absorbance vs. wavelength) kicking around if you want them.

**Comment by devbioinfoguy** [comment]
Greetings! I have a ton (ok, quite a #) of Excel files I've generated over the past 10 years doing evolution and developmental biology research. Happy to pass a few along.

Interestingly, DataCarpentry (sister group to Software Carpentry (SWC)) is looking to roll out workshops to subject matter areas: biology (genomics) and social sciences first. Structured data/Excel is going to be one of their modules. And its possible that SWC may talk about some of this in their SQL module (http://software-carpentry.org/lessons.html).

**Comment by posixeleni** [comment]
Thanks so much @pbulsink and @devbioinfoguy!
We would love any anonymized xslx files but @landreev can let you know if he needs anything specific to test with. Please email us your xlsx files to supportATdataverseDOTorg and if you dont mind we may bug you to confirm with us that these files are being ingested correctly.

**Comment by landreev** [comment]
@pbulsink 
@devbioinfoguy 
Philip, Bob,
Yeah, what Eleni said - thanks so much!
As long as they are XLSX files (i.e, the "new", xml-based Excel format - as opposed to the old, proprietary MS binary format)... We are not looking for any specific type of files; as in, we don't know yet what to specifically look for. Support for Excel format is a brand new feature in the Dataverse. Because, historically, Social Sciences have been (used to be) more of an SPSS/Stata/R shop... Getting data out of XLSX *seems* like a relatively straightforward thing - as long as it follows the basic pattern of a rectangular data matrix - variables (rows) and observations - we should be able to turn it into our data table... But, it's reasonable to expect that once we try the whole thing with files from real researchers, there will be some issues; some bad formatting, something stripped/lost, etc. So any real-life examples of real research data would be of much help. To try to ingest, to try to run our data exploration and analysis tools on, etc. 
And it sounds like what you have is as "real life" as it gets. 
Thanks again, 
-Leonid

**Comment by tracykteal** [comment]
People with data in Excel aren't usually thinking about exporting it to another tool and rarely have it formatted in a way that would make it easy or even possible to ingest. As @devbioinfoguy mentions, that's why in Data Carpentry, we have a 'data organization in spreadsheets' lesson (https://github.com/datacarpentry/datacarpentry/tree/master/lessons/excel/ecology-examples). So, it sounds like you're looking for Excel spreadsheets, from domain scientists, that would attempt to conform to the guidelines @posixeleni lays out, rather than just some standard wild Excel data? If you had a blog post on a call for this type of data, or whatever else you need, we'd be happy to post it at datacarpentry.org Thanks for all the work you're doing!

**Comment by posixeleni** [comment]
Ticket in RT w/ excel files for @landreev to test out:
> Queue: DVN_support
> Subject: Anonymized Excel Files
> Ticket URL: https://help.hmdc.harvard.edu/Ticket/Display.html?id=192570

**Comment by landreev** [comment]
I didn't have time to finish writing up an update for this, for beta 12. I'll get it done for 13, and I'll push in a couple of small ingest fixes I've added since then.

**Comment by eaquigley** [comment]
Based on comments in this ticket and additional research, @mcrosas will create requirements for this ingest.

**Comment by bencomp** [comment]
Related: #2626

**Comment by zward** [comment]
We use .csv files as a 'lowest common denominator' filetype for a project aimed largely at in-country partners.  They are well-structured/formatted (no spaces in header rows, no blanks, no mixing of datatypes in the same column, etc.) but still not compatible with TwoRavens.  One option is to upload them as Rdata files and let users download the tsv created during ingest, but ideally we would like to use our files as-is.  What about an option to add field-specific metadata during ingest that's not picked up from the file?

**Comment by pdurbin** [comment]
@tracykteal thanks for linking to https://github.com/datacarpentry/archive-datacarpentry/tree/master/lessons/excel/ecology-examples which seems like good advice for researchers using Excel.

@zward I'm not exactly sure what you mean by "field specific metadata" but in DVN 3.x we used to have the idea of a "control card" file (see http://guides.dataverse.org/en/3.6.2/dataverse-user-main.html#control-card-based-data-ingest ) but from what I understand we got away from this concept in Dataverse 4 because it was too confusing.

**Comment by pdurbin** [comment]
An another example of an xlsx file that failed ingest is  Harvard_Dataverse_homepage_survey_results_assessment.xlsx  from https://dataverse.harvard.edu/file.xhtml?fileId=2968911&version=1.1 so perhaps we can work with @jggautier on figuring how either how to improve ingest to accommodate that file or improve the docs to give users more help in formatting their file so that ingest can succeed.

**Comment by jggautier** [comment]
Just ran into a guide from Cornell about preparing tabular data for archiving: https://data.research.cornell.edu/content/tabular-data. (Looked but couldn't find anything similar at Harvard.) Could be helpful if we decided to improve the docs.

**Comment by mercecrosas** [comment]
This is a nice documentation piece, but it's written too much from only the
perspective of archiving, for what we need. I agree that the docs should
include this type of documentation on working with files, and files to
upload, but we should write from the point of view of a researcher or data
author - a researcher working with quantitative data usually will already
work with R, stata or similar data package. If he/she doesn't use any of
those, recommending what you should do when creating the CSV file can be
helpful.

Let's discuss this - it would be a very useful exercise to write this as
part of the docs.

Merce


----------
Mercè Crosas, Ph.D., Chief Data Science and Technology Officer, IQSS, Harvard
University
@mercecrosas mercecrosas.com

**Comment by bencomp** [comment]
Have you considered delegating CSV support to external libraries like [Apache Commons CSV](https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVParser.html)? I see it is listed as a dependency for Dataverse, but the ingest apparently still depends on @landreev's own code.

**Comment by pdurbin** [comment]
@bencomp I'm not sure why we invented our own way of parsing CSV files within the Dataverse code base. I agree with you that if we can leverage an open source package that already does this job, it would probably be better. Maybe there's some Dataverse-specific stuff in our CSV parsing. I don't know. @landreev would.

Yesterday @raprasad was showing us how Python's Pandas can parse Excel files: https://github.com/IQSS/miniverse/blob/729a8d1cb0bededbcc9882e51ccf5b874b1a178e/dv_apps/datafiles/tabular_previewer.py#L107

There's also a suggestion to have ingest run on a separate service rather than as part of the Dataverse monolith: #2331.

**Comment by landreev** [comment]
There was some historical reason why we insisted on one line per one observation requirement. But it doesn't appear to be relevant anymore. 
Also, when that original implementation was put together, we did not anticipate that anyone would use csv ingest for real, full text - with rich punctuation, extra tabs, commas and newlines. Since this appears to be a requirement now, yes, switching to an existing third party parser that takes care of all these things seems like the right thing to do.

**Comment by landreev** [comment]
I'm opening a new issue for some concrete incremental improvements of the CSV ingest, to be estimated and included into some sprint in the short term. (This one is just too generic)

**Comment by landreev** [comment]
#3767 - a child issue I created just for rewriting the CSV plugin to use more standard (and permissive) parsing.

**Comment by oscardssmith** [comment]
Does #3767  close this?

**Comment by adam3smith** [comment]
I have two additions on this -- there are a couple of tickets on Excel ingest, this seemed the most appropriate one but please redirect if I missed something
1. .tab vs. .tsv: I like that Dataverse normalized .xlsx files and I think tab separated values are a good, robust choice, but why did you choose the .tab file extension (which is ambiguous and very rarely used) over the much more common .tsv? 
2. The download dialog for ingested tabular files has an option to download the "original" file. I think that should be the original xlsx file, as uploaded. Currently it's just the .tab file again.

**Comment by pdurbin** [comment]
@adam3smith thanks, these are related:

- Switch preservation format extension from .tab to .tsv #2720
- Default Download Type should be "original" #4000

**Comment by adam3smith** [comment]
Great. #2720 addresses what I want with 1. but while I agree with Alex in #4000 , my point about Excel is different: the original format is completely lost when uploading Excel. I have no way to get out the Excel file I uploaded after ingest (at least I'm not finding it). That may actually be a bug?

**Comment by jggautier** [comment]
HI @adam3smith. The second point sounds like a bug. Could you link to a dataset with an ingested Excel file where we're not able to download the original format?

This quick search returns a few ingested Excel files, but I'm able to download the original format: https://dataverse.harvard.edu/dataverse/harvard?q=fileName%3Aexcel&fq1=publicationStatus%3A%22Published%22&fq0=fileTypeGroupFacet%3A%22tabulardata%22&types=dataverses%3Adatasets%3Afiles&sort=score&order=desc

**Comment by adam3smith** [comment]
Sorry for the false alarm -- this works correctly for me not just in demo.dataverse but also in our dev instance on 4.8. It seems broken the way I describe, though, in our beta environment running 4.6.2. So may have gotten fixed at some point?

**Comment by mheppler** [comment]
@adam3smith there were a lot of ingest improvements included in our [4.8 release](https://github.com/IQSS/dataverse/releases/tag/v4.8), which could account for improved functionality between demo (4.8.4) and your 4.6.2.

**Comment by landreev** [comment]
Since this issue was opened, it has been addressed for the CSV ingest plugin, that was much improved. 
I propose we close this issue, but open a new separate issue for the Excel plugin. And try to address it asap (there is a bunch of open Excel issues for various things the current plugin does wrong)

**Comment by pdurbin** [comment]
@landreev yeah, I agree that this issue should probably be closed. It's one of our oldest open issues and doesn't fit into our modern "small chunks" style. #3382 ("Excel ingest can't accept more than 26 (Z) columns") is an example of a small chunk and @qqmyers make pull request #5891 for it yesterday. (Thank you!! 🎉 ) Then @landreev opened #5896 about adding tests for Excel. This is perfect. Small chunks move across our (new) board at https://github.com/orgs/IQSS/projects/2 much faster than large projects.

@adam3smith you seem to be fairly engaged in this issue. Are you willing or able to create a "small chunk" issue or two for some of your ideas above?

Everyone else is very welcome to create new issues as well, of course! The idea is that the issues should be relatively small and targeted. Incremental improvement. Small chunks.

**Comment by kcondon** [comment]
@pdurbin @landreev I opened this issue originally because the initial implementation was very basic because we did not have a good idea of what a set of features we should support would be, given the wide variety of practices. It was, at least initially, a research spike to figure that out, then become something specific. We lucked out there was a csv implementation so I agree that part is resolved. I think Leonid's suggestion of opening a new Excel-specific implementation sounds reasonable and I'm happy it has specifics since when I opened this there was not an agreed upon set of specifics. So, 1. open new ticket with specifics, then 2. close this ticket ;)
